compact_str/src/to_compact_str_specialisation.rs

```diff
@@ -0,0 +1,164 @@
+use super::CompactStr;
+
+use castaway::{cast, LifetimeFree};
+
+/// # Safety
+///
+///  - CompactStr does not contain any lifetime
+///  - CompactStr is 'static
+///  - CompactStr is a container to `u8`, which is `LifetimeFree`.
+unsafe impl LifetimeFree for CompactStr {}
+
+const TRUE_COMPACT_STR: CompactStr = CompactStr::new_inline("true");
+const FALSE_COMPACT_STR: CompactStr = CompactStr::new_inline("false");
+
+#[inline(always)]
+pub(super) fn to_compact_str_specialised<T>(val: &T) -> Option<CompactStr> {
```

Comment:
I'll need to take a closer look at this to figure out how it's actually zero cost, right now there seem to be some runtime checks?

Out of curiosity, is it not possible to do autoref specialization here?

Reply:
According to @sagebind , all of these cast! can be optimized out by compiler.

I briefly read through the code and I also think it is possible for compiler to prune all the unused specialisation.

Reply:
"runtime checks" is a bit hard to define in light of optimization; technically yes, Castaway macros generate program code to make appropriate checks. But these checks are trivial from the compiler's perspective (think `if size_of::<A>() == size_of::<B>()`) and are always reduced to a simple `true` or `false` at even just a basic optimization level; once this happens it is also trivial for the compiler to eliminate a branch in _your_ code, since it effectively looks like `if true {} else {}` (or `if false {} else {}` depending on which type it is called with) to rustc by the time that stage of compilation is reached.

> Out of curiosity, is it not possible to do autoref specialization here?

Nope, auto(de)ref techniques don't work inside a generic context, because ref specialization is resolved _before_ monomorphization happens. (Which is a good thing by the way, otherwise you'd have C++ problems, such as compilation failing "sometimes" depending on which concrete types you supply to a template.)

Reply:
@ParkMyCar [Here is the code](https://github.com/sagebind/castaway/blob/52987d6ebbdfcc451fb24585dcb34590f6e62657/src/utils.rs#L29) responsible for checking whether two non-`' static` types are the same, which is at the core of `castaway`.

I am pretty sure they can be easily evaluated at compile time.

Reply:
> since it effectively looks like `if true {} else {}` (or `if false {} else {}`)
That's wild, super cool work @sagebind!